testnet/stacks-node/src/neon_node.rs

```diff
@@ -139,2179 +483,3505 @@ fn fault_injection_long_tenure() {
-                Some(k) => k,
-                None => {
-                    error!(
-                        "Failed to obtain microblock key for mining attempt";
-                        "attempt" => %attempt
-                    );
-                    return None;
-                }
+        if !self.miner_thread_try_join() {
+            // already running (for an anchored block or microblock)
+            debug!("Relayer: miner thread already running so cannot mine microblock");
+            return false;
+        }
+        if self
+            .globals
+            .get_miner_status()
+            .lock()
+            .expect("FATAL: mutex poisoned")
+            .is_blocked()
+        {
+            debug!(
+                "Relayer: miner is blocked as of {}; cannot mine microblock at this time",
+                &burnchain_tip.burn_header_hash
+            );
+            self.globals.counters.set_microblocks_processed(0);
+            return false;
+        }
+
+        let parent_consensus_hash = &miner_tip.consensus_hash;
+        let parent_block_hash = &miner_tip.block_hash;
+
+        debug!(
+            "Relayer: Run microblock tenure for {}/{}",
+            parent_consensus_hash, parent_block_hash
+        );
+
+        let mut microblock_thread_state = match MicroblockMinerThread::from_relayer_thread(self) {
+            Some(ts) => ts,
+            None => {
+                return false;
             }
-        } else {
-            keychain.rotate_microblock_keypair(burn_block.block_height)
         };
-        let mblock_pubkey_hash =
-            Hash160::from_node_public_key(&StacksPublicKey::from_private(&microblock_secret_key));
 
-        let coinbase_tx = inner_generate_coinbase_tx(
-            keychain,
-            coinbase_nonce,
-            config.is_mainnet(),
-            config.burnchain.chain_id,
-        );
+        if let Ok(miner_handle) = thread::Builder::new()
+            .name(format!("miner-microblock-{}", self.local_peer.data_url))
+            .spawn(move || {
+                Some(MinerThreadResult::Microblock(
+                    microblock_thread_state.try_mine_microblock(miner_tip.clone()),
+                    miner_tip,
+                ))
+            })
+            .map_err(|e| {
+                error!("Relayer: Failed to start tenure thread: {:?}", &e);
+                e
+            })
+        {
+            // thread started!
+            self.miner_thread = Some(miner_handle);
+            self.microblock_deadline =
+                get_epoch_time_ms() + (self.config.node.microblock_frequency as u128);
+        }
+
+        true
+    }
+
+    /// Inner body of Self::miner_thread_try_join
+    fn inner_miner_thread_try_join(
+        &mut self,
+        thread_handle: JoinHandle<Option<MinerThreadResult>>,
+    ) -> Option<JoinHandle<Option<MinerThreadResult>>> {
+        // tenure run already in progress; try and join
+        if !thread_handle.is_finished() {
```

Comment:
Quick note from trying to start the mock-miner: I ran into a compile-time error here because of an older `rustc` version that didn't support this method. Should we add that requirement into the Cargo.toml?

```toml
rust-version = "1.61"
```

Reply:
yup, saw that too.  Will fix.